From 85a0aa4df0b0005c7c8f972262db0f3da860408d Mon Sep 17 00:00:00 2001
From: Antonin Godard <antoningodard@pm.me>
Date: Thu, 15 Aug 2024 18:48:10 -0700
Subject: [PATCH] Add inactive_focused_workspace support

This new i3bar configuration allows a user to configure the color,
border and background of a workspace button that is currently inactive
_and_ focused - "focused" meaning that the workspace is currently under
the focused monitor.

Unlike previously, this makes its configuration distinct from the
inactive_workspace button. It is especially useful when the bar color is
different on a focused and unfocused monitor.

By default the inactive_focused_workspace colors remain the same as the
inactive_workspace colors, so the change won't be notived unless
explicitely configured.

Signed-off-by: Antonin Godard <antoningodard@pm.me>
---
 docs/ipc                     |  3 +++
 docs/userguide               | 14 +++++++++-----
 i3bar/include/xcb.h          |  3 +++
 i3bar/src/config.c           |  6 ++++++
 i3bar/src/xcb.c              | 11 +++++++++++
 include/configuration.h      |  4 ++++
 parser-specs/config.spec     |  2 +-
 src/config.c                 |  3 +++
 src/config_directives.c      |  1 +
 src/ipc.c                    |  3 +++
 testcases/t/177-bar-config.t | 25 ++++++++++++++++---------
 11 files changed, 60 insertions(+), 15 deletions(-)

diff --git a/docs/ipc b/docs/ipc
index 8d8bdfa5..d70a3be2 100644
--- a/docs/ipc
+++ b/docs/ipc
@@ -702,6 +702,9 @@ inactive_workspace_text/inactive_workspace_bg/inactive_workspace_border::
 	Text/background/border color for a workspace button when the workspace
 	does not have focus and is not active (visible) on any output. This
 	will be the case for most workspaces.
+inactive_focused_workspace_text/inactive_focused_workspace_bg/inactive_focused_workspace_border::
+	Text/background/border color for a workspace button when the workspace
+	is not in focus but in the focused monitor.
 urgent_workspace_text/urgent_workspace_bg/urgent_workspace_border::
 	Text/background/border color for workspaces which contain at least one
 	window with the urgency hint set.
diff --git a/docs/userguide b/docs/userguide
index d211c1d5..4e5b5349 100644
--- a/docs/userguide
+++ b/docs/userguide
@@ -2056,6 +2056,9 @@ inactive_workspace::
 	Border, background and text color for a workspace button when the workspace
 	does not have focus and is not active (visible) on any output. This
 	will be the case for most workspaces.
+inactive_focused_workspace::
+	Border, background and text color for a workspace button when the workspace
+	is not focused but is part of the monitor in focus.
 urgent_workspace::
 	Border, background and text color for a workspace button when the workspace
 	contains a window with the urgency hint set.
@@ -2082,11 +2085,12 @@ bar {
         statusline #ffffff
         separator #666666
 
-        focused_workspace  #4c7899 #285577 #ffffff
-        active_workspace   #333333 #5f676a #ffffff
-        inactive_workspace #333333 #222222 #888888
-        urgent_workspace   #2f343a #900000 #ffffff
-        binding_mode       #2f343a #900000 #ffffff
+        focused_workspace          #4c7899 #285577 #ffffff
+        active_workspace           #333333 #5f676a #ffffff
+        inactive_workspace         #333333 #222222 #888888
+        inactive_focused_workspace #333333 #222222 #888888
+        urgent_workspace           #2f343a #900000 #ffffff
+        binding_mode               #2f343a #900000 #ffffff
     }
 }
 --------------------------------------
diff --git a/i3bar/include/xcb.h b/i3bar/include/xcb.h
index fa274d01..50e42669 100644
--- a/i3bar/include/xcb.h
+++ b/i3bar/include/xcb.h
@@ -41,6 +41,9 @@ struct xcb_color_strings_t {
     char *focus_ws_bg;
     char *focus_ws_fg;
     char *focus_ws_border;
+    char *inactive_focus_ws_bg;
+    char *inactive_focus_ws_fg;
+    char *inactive_focus_ws_border;
     char *urgent_ws_bg;
     char *urgent_ws_fg;
     char *urgent_ws_border;
diff --git a/i3bar/src/config.c b/i3bar/src/config.c
index cebd5d5d..08b95d54 100644
--- a/i3bar/src/config.c
+++ b/i3bar/src/config.c
@@ -256,6 +256,9 @@ static int config_string_cb(void *params_, const unsigned char *val, size_t _len
     COLOR(inactive_workspace_border, inactive_ws_border);
     COLOR(inactive_workspace_bg, inactive_ws_bg);
     COLOR(inactive_workspace_text, inactive_ws_fg);
+    COLOR(inactive_focused_workspace_border, inactive_focus_ws_border);
+    COLOR(inactive_focused_workspace_bg, inactive_focus_ws_bg);
+    COLOR(inactive_focused_workspace_text, inactive_focus_ws_fg);
     COLOR(urgent_workspace_border, urgent_ws_border);
     COLOR(urgent_workspace_bg, urgent_ws_bg);
     COLOR(urgent_workspace_text, urgent_ws_fg);
@@ -477,6 +480,9 @@ void free_colors(struct xcb_color_strings_t *colors) {
     FREE_COLOR(focus_ws_fg);
     FREE_COLOR(focus_ws_bg);
     FREE_COLOR(focus_ws_border);
+    FREE_COLOR(inactive_focus_ws_fg);
+    FREE_COLOR(inactive_focus_ws_bg);
+    FREE_COLOR(inactive_focus_ws_border);
     FREE_COLOR(binding_mode_fg);
     FREE_COLOR(binding_mode_bg);
     FREE_COLOR(binding_mode_border);
diff --git a/i3bar/src/xcb.c b/i3bar/src/xcb.c
index 4ff44d27..db832ffb 100644
--- a/i3bar/src/xcb.c
+++ b/i3bar/src/xcb.c
@@ -100,6 +100,9 @@ struct xcb_colors_t {
     color_t focus_ws_bg;
     color_t focus_ws_fg;
     color_t focus_ws_border;
+    color_t inactive_focus_ws_bg;
+    color_t inactive_focus_ws_fg;
+    color_t inactive_focus_ws_border;
     color_t binding_mode_bg;
     color_t binding_mode_fg;
     color_t binding_mode_border;
@@ -407,6 +410,9 @@ void init_colors(const struct xcb_color_strings_t *new_colors) {
     PARSE_COLOR(focus_ws_fg, "#FFFFFF");
     PARSE_COLOR(focus_ws_bg, "#285577");
     PARSE_COLOR(focus_ws_border, "#4c7899");
+    PARSE_COLOR(inactive_focus_ws_fg, "#888888");
+    PARSE_COLOR(inactive_focus_ws_bg, "#222222");
+    PARSE_COLOR(inactive_focus_ws_border, "#333333");
 #undef PARSE_COLOR
 
 #define PARSE_COLOR_FALLBACK(name, fallback)                                                         \
@@ -2054,6 +2060,7 @@ void draw_bars(bool unhide) {
                 color_t fg_color = colors.inactive_ws_fg;
                 color_t bg_color = colors.inactive_ws_bg;
                 color_t border_color = colors.inactive_ws_border;
+                bool ws_output_has_focus = output_has_focus(ws_walk->output);
                 if (ws_walk->visible) {
                     if (!ws_walk->focused) {
                         fg_color = colors.active_ws_fg;
@@ -2064,6 +2071,10 @@ void draw_bars(bool unhide) {
                         bg_color = colors.focus_ws_bg;
                         border_color = colors.focus_ws_border;
                     }
+                } else if (ws_output_has_focus) {
+                    fg_color = colors.inactive_focus_ws_fg;
+                    bg_color = colors.inactive_focus_ws_bg;
+                    border_color = colors.inactive_focus_ws_border;
                 }
                 if (ws_walk->urgent) {
                     DLOG("WS %s is urgent!\n", i3string_as_utf8(ws_walk->name));
diff --git a/include/configuration.h b/include/configuration.h
index 19d2f714..142cfd4f 100644
--- a/include/configuration.h
+++ b/include/configuration.h
@@ -394,6 +394,10 @@ struct Barconfig {
         char *inactive_workspace_bg;
         char *inactive_workspace_text;
 
+        char *inactive_focused_workspace_border;
+        char *inactive_focused_workspace_bg;
+        char *inactive_focused_workspace_text;
+
         char *urgent_workspace_border;
         char *urgent_workspace_bg;
         char *urgent_workspace_text;
diff --git a/parser-specs/config.spec b/parser-specs/config.spec
index 33708b52..74bec885 100644
--- a/parser-specs/config.spec
+++ b/parser-specs/config.spec
@@ -728,7 +728,7 @@ state BAR_COLORS:
   'set' -> BAR_COLORS_IGNORE_LINE
   colorclass = 'background', 'statusline', 'separator', 'focused_background', 'focused_statusline', 'focused_separator'
       -> BAR_COLORS_SINGLE
-  colorclass = 'focused_workspace', 'active_workspace', 'inactive_workspace', 'urgent_workspace', 'binding_mode'
+  colorclass = 'focused_workspace', 'active_workspace', 'inactive_workspace', 'inactive_focused_workspace', 'urgent_workspace', 'binding_mode'
       -> BAR_COLORS_BORDER
   '}'
       -> BAR
diff --git a/src/config.c b/src/config.c
index bf3ec6dc..9c8d4bec 100644
--- a/src/config.c
+++ b/src/config.c
@@ -123,6 +123,9 @@ static void free_configuration(void) {
         FREE(barconfig->colors.inactive_workspace_border);
         FREE(barconfig->colors.inactive_workspace_bg);
         FREE(barconfig->colors.inactive_workspace_text);
+        FREE(barconfig->colors.inactive_focused_workspace_border);
+        FREE(barconfig->colors.inactive_focused_workspace_bg);
+        FREE(barconfig->colors.inactive_focused_workspace_text);
         FREE(barconfig->colors.urgent_workspace_border);
         FREE(barconfig->colors.urgent_workspace_bg);
         FREE(barconfig->colors.urgent_workspace_text);
diff --git a/src/config_directives.c b/src/config_directives.c
index 81adf351..02404d20 100644
--- a/src/config_directives.c
+++ b/src/config_directives.c
@@ -832,6 +832,7 @@ CFGFUN(bar_color, const char *colorclass, const char *border, const char *backgr
     APPLY_COLORS(focused_workspace);
     APPLY_COLORS(active_workspace);
     APPLY_COLORS(inactive_workspace);
+    APPLY_COLORS(inactive_focused_workspace);
     APPLY_COLORS(urgent_workspace);
     APPLY_COLORS(binding_mode);
 
diff --git a/src/ipc.c b/src/ipc.c
index d20090c6..a74f7d44 100644
--- a/src/ipc.c
+++ b/src/ipc.c
@@ -886,6 +886,9 @@ static void dump_bar_config(yajl_gen gen, Barconfig *config) {
     YSTR_IF_SET(inactive_workspace_border);
     YSTR_IF_SET(inactive_workspace_bg);
     YSTR_IF_SET(inactive_workspace_text);
+    YSTR_IF_SET(inactive_focused_workspace_border);
+    YSTR_IF_SET(inactive_focused_workspace_bg);
+    YSTR_IF_SET(inactive_focused_workspace_text);
     YSTR_IF_SET(urgent_workspace_border);
     YSTR_IF_SET(urgent_workspace_bg);
     YSTR_IF_SET(urgent_workspace_text);
diff --git a/testcases/t/177-bar-config.t b/testcases/t/177-bar-config.t
index 1ced3ce0..baef7be4 100644
--- a/testcases/t/177-bar-config.t
+++ b/testcases/t/177-bar-config.t
@@ -115,11 +115,12 @@ bar {
         focused_statusline #cccc00
         focused_separator  #0000cc
 
-        focused_workspace   #4c7899 #285577 #ffffff
-        active_workspace    #333333 #222222 #888888
-        inactive_workspace  #333333 #222222 #888888
-        urgent_workspace    #2f343a #900000 #ffffff
-        binding_mode        #abc123 #123abc #ababab
+        focused_workspace           #4c7899 #285577 #ffffff
+        active_workspace            #333333 #222222 #888888
+        inactive_workspace          #333333 #222222 #888888
+        inactive_focused_workspace  #333333 #222222 #888888
+        urgent_workspace            #2f343a #900000 #ffffff
+        binding_mode                #abc123 #123abc #ababab
     }
 
     # old syntax for compatibility with i3-gaps
@@ -175,6 +176,9 @@ is_deeply($bar_config->{colors},
         inactive_workspace_border => '#333333',
         inactive_workspace_text => '#888888',
         inactive_workspace_bg => '#222222',
+        inactive_focused_workspace_border => '#333333',
+        inactive_focused_workspace_text => '#888888',
+        inactive_focused_workspace_bg => '#222222',
         urgent_workspace_border => '#2f343a',
         urgent_workspace_text => '#ffffff',
         urgent_workspace_bg => '#900000',
@@ -385,10 +389,11 @@ bar {
         background #ff0000
         statusline   #00ff00
 
-        focused_workspace   #ffffff #285577
-        active_workspace    #888888 #222222
-        inactive_workspace  #888888 #222222
-        urgent_workspace    #ffffff #900000
+        focused_workspace          #ffffff #285577
+        active_workspace           #888888 #222222
+        inactive_workspace         #888888 #222222
+        inactive_focused_workspace #888888 #222222
+        urgent_workspace           #ffffff #900000
     }
 }
 EOT
@@ -423,6 +428,8 @@ is_deeply($bar_config->{colors},
         active_workspace_bg => '#222222',
         inactive_workspace_text => '#888888',
         inactive_workspace_bg => '#222222',
+        inactive_focused_workspace_text => '#888888',
+        inactive_focused_workspace_bg => '#222222',
         urgent_workspace_text => '#ffffff',
         urgent_workspace_bg => '#900000',
     }, '(old) colors ok');
-- 
2.46.0

